Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php

```diff
@@ -87,6 +90,51 @@ static function _checkGoto($a_target)
 
 		return false;
 	}
+
+	/**
+	 * Get operators
+	 */
+	public static function getConditionOperators()
+	{
+		include_once './Services/AccessControl/classes/class.ilConditionHandler.php';
+		return array(
+			ilConditionHandler::OPERATOR_ACCREDITED_OR_PASSED
+		);
+	}
+
+	/**
+	 *
+	 * @global ilObjUser $ilUser
+	 * @param type $a_obj_id
+	 * @param type $a_operator
+	 * @param type $a_value
+	 * @param type $a_usr_id
+	 * @return boolean
+	 */
+	public static function checkCondition($a_obj_id,$a_operator,$a_value,$a_usr_id)
+	{
+		include_once './Services/AccessControl/classes/class.ilConditionHandler.php';
+		include_once "./Modules/StudyProgramme/classes/class.ilStudyProgrammeUserProgress.php";
+
+		switch($a_operator) {
+			case ilConditionHandler::OPERATOR_ACCREDITED_OR_PASSED:
+				$valid_progress = array(
+					ilStudyProgrammeProgress::STATUS_COMPLETED,
+					ilStudyProgrammeProgress::STATUS_ACCREDITED
+				);
+
+				$prg_user_progress = ilStudyProgrammeUserProgress::getInstancesForUser($a_obj_id, $a_usr_id);
+				foreach ($prg_user_progress as $progress) {
+					if( in_array($progress->getStatus(), $valid_progress)) {
+						return true;
+					}
+				}
```

Comment:
A note, about performance: this method is called for every 'visible' and 'read' permission check of SPs with preconditions. Maybe a caching would improve performance. Additionally one could use ilObjectAccess::_preloadData($obj_ids, $ref_ids) to load required data (for _checkCondition, _checkAccess) from the database in one step.